## 1. Importing the Libraries

In [58]:
import pandas as pd
import numpy as np

## 2. Loading the dataset and basic insepection

In [59]:
df = pd.read_csv("food_delivery_orders_dataset.csv")
df.head(10)

,order_id,customer_id,restaurant_id,driver_id,order_timestamp,order_date,order_hour,day_of_week,is_weekend,city,...,traffic_level,delivery_partner_experience_months,delivery_partner_rating,restaurant_preparation_time_minutes,estimated_delivery_time_minutes,actual_delivery_time_minutes,late_delivery,order_status,cancellation_reason,customer_rating
0,ORD_000001,CUS_00001,RES_0008,DRV_0075,2025-10-05 15:54:00,2025-10-05,15,Sunday,1,City_C,...,Medium,14,4.4,25,25,29.0,0.0,Completed,NaN,4.6
1,ORD_000002,CUS_00013,RES_0024,DRV_0131,2025-10-01 20:00:00,2025-10-01,20,Wednesday,0,City_A,...,High,44,4.7,12,20,24.0,0.0,Completed,NaN,5.0
2,ORD_000003,CUS_00049,RES_0013,DRV_0087,2025-10-07 18:24:00,2025-10-07,18,Tuesday,0,City_C,...,High,1,5.0,11,63,72.0,1.0,Completed,NaN,3.6
3,ORD_000004,CUS_00007,RES_0001,DRV_0177,2025-10-30 16:54:00,2025-10-30,16,Thursday,0,City_A,...,Low,50,4.3,12,23,31.0,1.0,Completed,NaN,4.1
4,ORD_000005,CUS_00325,RES_0029,DRV_0114,2025-10-15 18:34:00,2025-10-15,18,Wednesday,0,City_B,...,Medium,21,4.0,9,32,42.0,1.0,Completed,NaN,3.9
5,ORD_000006,CUS_00580,RES_0019,DRV_0112,2025-10-23 14:27:00,2025-10-23,14,Thursday,0,City_B,...,Medium,34,5.0,12,32,39.0,1.0,Completed,NaN,4.0
6,ORD_000007,CUS_00021,RES_0002,DRV_0109,2025-10-29 19:20:00,2025-10-29,19,Wednesday,0,City_D,...,Medium,43,4.8,11,27,31.0,0.0,Completed,NaN,4.5
7,ORD_000008,CUS_00003,RES_0044,DRV_0066,2025-10-26 07:22:00,2025-10-26,7,Sunday,1,City_D,...,Low,53,4.0,12,20,22.0,0.0,Completed,NaN,4.8
8,ORD_000009,CUS_00002,RES_0003,DRV_0155,2025-10-24 14:28:00,2025-10-24,14,Friday,0,City_C,...,High,46,4.5,11,37,40.0,0.0,Completed,NaN,5.0
9,ORD_000010,CUS_00009,RES_0009,DRV_0126,2025-10-14 20:32:00,2025-10-14,20,Tuesday,0,City_B,...,Medium,47,4.0,14,25,22.0,0.0,Completed,NaN,4.8


In [60]:
df.shape

(50000, 37)

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 37 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   order_id                             50000 non-null  object 
 1   customer_id                          50000 non-null  object 
 2   restaurant_id                        50000 non-null  object 
 3   driver_id                            50000 non-null  object 
 4   order_timestamp                      50000 non-null  object 
 5   order_date                           50000 non-null  object 
 6   order_hour                           50000 non-null  int64  
 7   day_of_week                          50000 non-null  object 
 8   is_weekend                           50000 non-null  int64  
 9   city                                 50000 non-null  object 
 10  delivery_area                        50000 non-null  object 
 11  customer_age                

In [62]:
df.describe()

,order_hour,is_weekend,customer_age,restaurant_rating,items_count,subtotal,discount_percent,tax_amount,service_fee,delivery_fee,order_total,tip_amount,distance_km,delivery_partner_experience_months,delivery_partner_rating,restaurant_preparation_time_minutes,estimated_delivery_time_minutes,actual_delivery_time_minutes,late_delivery,customer_rating
count,50000.000000,50000.00000,50000.000000,50000.000000,50000.00000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,49132.000000,50000.000000,50000.0000,50000.000000,50000.000000,50000.000000,49132.000000,49132.000000,49132.000000
mean,14.465900,0.26712,27.686580,4.207956,2.91002,30.897938,4.495700,4.128796,1.474600,3.075108,38.169501,1.610076,4.730632,29.7332,4.495062,15.549600,30.950580,35.152365,0.365709,4.295946
std,5.231155,0.44246,7.537796,0.450977,2.02907,27.068965,6.430008,3.636275,1.378101,2.299115,31.054669,2.878464,2.780190,17.4052,0.349198,7.360099,12.825519,14.701154,0.481633,0.530947
min,0.000000,0.00000,18.000000,2.800000,1.00000,5.100000,0.000000,0.570000,0.130000,0.000000,4.870000,0.000000,0.800000,1.0000,3.200000,5.000000,8.000000,10.000000,0.000000,1.800000
25%,11.000000,0.00000,21.000000,3.900000,1.00000,13.100000,0.000000,1.730000,0.580000,0.000000,17.430000,0.000000,2.800000,14.0000,4.300000,10.000000,22.000000,25.000000,0.000000,4.000000
50%,14.000000,0.00000,27.000000,4.200000,2.00000,22.825000,0.000000,2.890000,1.040000,3.500000,28.350000,0.000000,4.100000,29.0000,4.500000,14.000000,28.000000,33.000000,0.000000,4.400000
75%,19.000000,1.00000,33.000000,4.600000,4.00000,40.120000,10.000000,5.390000,1.890000,4.500000,48.660000,2.280000,5.900000,44.0000,4.800000,19.000000,37.000000,42.000000,1.000000,4.700000
max,23.000000,1.00000,55.000000,5.000000,10.00000,362.980000,20.000000,50.820000,25.210000,12.000000,443.210000,53.650000,20.000000,60.0000,5.000000,73.000000,128.000000,172.000000,1.000000,5.000000


## 3. Checking columns

In [63]:
df["city"].value_counts()

city
City_A    19888
City_B    14885
City_C    10178
City_D     5049
Name: count, dtype: int64

In [64]:
df["order_status"].value_counts()

order_status
Completed    49132
Cancelled      868
Name: count, dtype: int64

### Duplicate check

In [65]:
df["order_id"].duplicated().sum()


np.int64(0)

In [66]:
df.duplicated().sum()

np.int64(0)

## 4. Missing values
Find which columns actually have nulls, and how many.

In [67]:
null_value_columns = []
for i in df.columns:
    if df[i].isnull().sum() > 0:
        null_value_columns.append(i)
    else:
        continue

null_value_columns


['tip_amount',
 'actual_delivery_time_minutes',
 'late_delivery',
 'cancellation_reason',
 'customer_rating']

In [68]:
df[null_value_columns].isnull().sum()

tip_amount                        868
actual_delivery_time_minutes      868
late_delivery                     868
cancellation_reason             49132
customer_rating                   868
dtype: int64

As we can see the cancellation_reason column has 49132 null values (NO Reason) out of 50,000. which simply means the 49132 orders aren't Cancelled they have deliverd successfully. 
Also not to mention there are 868 null values in late_delivery,Customer_Rating and Actual_delivery_time_minutes columns, which indicates these value comes from the delivery that got cancelled because 49132 + 868 = 50,000 which is the actual length of the dataset. Hence there are no orders with tip amount, late delivery, customer rating. actual delivery timme minutes that has delivered.



## 5. Cleaning

In [69]:
df["cancellation_reason"] = df["cancellation_reason"].fillna("Not Cancelled")
df["cancellation_reason"].value_counts()

cancellation_reason
Not Cancelled             49132
Driver Unassigned           386
Restaurant Unavailable      160
Customer Cancelled          154
Changed Mind                153
Long Prep Delay              15
Name: count, dtype: int64

In [70]:
# A cancelled order was never tipped
df["tip_amount"] = df["tip_amount"].fillna(0)

In [71]:
# No delivery happened, so there's no real delivery time or lateness flag to impute.
# Filling with the average of completed orders keeps the column numeric for Power BI
# without pretending the order was on time.
df["actual_delivery_time_minutes"] = df["actual_delivery_time_minutes"].fillna(df["actual_delivery_time_minutes"].mean())
df["late_delivery"] = df["late_delivery"].fillna(df["late_delivery"].mean())

In [72]:
# No customer rating was ever given for a cancelled order
df["customer_rating"] = df["customer_rating"].fillna(df["customer_rating"].mean())

### Confirming nothing is missing anymore

In [73]:
df.isnull().sum().sum()

np.int64(0)

## 6. Fix data types
`order_date` and `order_timestamp` are read in as text, converting them to real dates makes them usable for time-based visuals in Power BI

In [74]:
df["order_date"] = pd.to_datetime(df["order_date"])
df["order_timestamp"] = pd.to_datetime(df["order_timestamp"])
df.dtypes[["order_date", "order_timestamp"]]

order_date         datetime64[ns]
order_timestamp    datetime64[ns]
dtype: object

## 7. Final check

In [75]:
print("Shape:", df.shape)
print("Total nulls:", df.isnull().sum().sum())
df.head()

Shape: (50000, 37)
Total nulls: 0


,order_id,customer_id,restaurant_id,driver_id,order_timestamp,order_date,order_hour,day_of_week,is_weekend,city,...,traffic_level,delivery_partner_experience_months,delivery_partner_rating,restaurant_preparation_time_minutes,estimated_delivery_time_minutes,actual_delivery_time_minutes,late_delivery,order_status,cancellation_reason,customer_rating
0,ORD_000001,CUS_00001,RES_0008,DRV_0075,2025-10-05 15:54:00,2025-10-05,15,Sunday,1,City_C,...,Medium,14,4.4,25,25,29.0,0.0,Completed,Not Cancelled,4.6
1,ORD_000002,CUS_00013,RES_0024,DRV_0131,2025-10-01 20:00:00,2025-10-01,20,Wednesday,0,City_A,...,High,44,4.7,12,20,24.0,0.0,Completed,Not Cancelled,5.0
2,ORD_000003,CUS_00049,RES_0013,DRV_0087,2025-10-07 18:24:00,2025-10-07,18,Tuesday,0,City_C,...,High,1,5.0,11,63,72.0,1.0,Completed,Not Cancelled,3.6
3,ORD_000004,CUS_00007,RES_0001,DRV_0177,2025-10-30 16:54:00,2025-10-30,16,Thursday,0,City_A,...,Low,50,4.3,12,23,31.0,1.0,Completed,Not Cancelled,4.1
4,ORD_000005,CUS_00325,RES_0029,DRV_0114,2025-10-15 18:34:00,2025-10-15,18,Wednesday,0,City_B,...,Medium,21,4.0,9,32,42.0,1.0,Completed,Not Cancelled,3.9


## 8. Export for Power BI

In [ ]:
df.to_csv("new_food_delivery_orders_cleaned.csv", index=False)